internal/transport/grpchttp2/http2bridge.go

```diff
@@ -0,0 +1,214 @@
+	"golang.org/x/net/http2/hpack"
+	"google.golang.org/grpc"
+)
+
+const (
+	initHeaderTableSize = 4096
+	maxFrameLen         = 16384
+)
+
+type HTTP2FramerBridge struct {
+	framer *http2.Framer
+	buf    [maxFrameLen]byte
+	pool   grpc.SharedBufferPool
+}
+
+func NewHTTP2FramerBridge(w io.Writer, r io.Reader, maxHeaderListSize uint32) *HTTP2FramerBridge {
+	fr := &HTTP2FramerBridge{
+		framer: http2.NewFramer(w, r),
+		pool:   grpc.NewSharedBufferPool(),
+	}
+
+	fr.framer.SetReuseFrames()
+	fr.framer.MaxHeaderListSize = maxHeaderListSize
+	fr.framer.ReadMetaHeaders = (hpack.NewDecoder(initHeaderTableSize, nil))
+
+	return fr
+}
+
+func (fr *HTTP2FramerBridge) ReadFrame() (Frame, error) {
+	f, err := fr.framer.ReadFrame()
+	if err != nil {
+		return nil, err
+	}
+
+	h := f.Header()
+	hdr := &FrameHeader{
+		Size:     h.Length,
+		Type:     FrameType(h.Type),
+		Flags:    Flag(h.Flags),
+		StreamID: h.StreamID,
+	}
+
+	switch f := f.(type) {
+	case *http2.DataFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.Data())
+		df := &DataFrame{
+			hdr:  hdr,
+			Data: buf,
+		}
+		df.free = func() {
+			fr.pool.Put(&buf)
+			df.Data = nil
+		}
+		return df, nil
+	case *http2.HeadersFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.HeaderBlockFragment())
+		hf := &HeadersFrame{
+			hdr:      hdr,
+			HdrBlock: buf,
+		}
+		hf.free = func() {
+			fr.pool.Put(&buf)
+			hf.HdrBlock = nil
+		}
+		return hf, nil
+	case *http2.RSTStreamFrame:
+		return &RSTStreamFrame{
+			hdr:  hdr,
+			Code: ErrCode(f.ErrCode),
+		}, nil
+	case *http2.SettingsFrame:
+		buf := make([]Setting, 0, f.NumSettings())
+		f.ForeachSetting(func(s http2.Setting) error {
+			buf = append(buf, Setting{
+				ID:    SettingID(s.ID),
+				Value: s.Val,
+			})
+			return nil
+		})
+		sf := &SettingsFrame{
+			hdr:      hdr,
+			Settings: buf,
+		}
+		return sf, nil
+	case *http2.PingFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.Data[:])
+		pf := &PingFrame{
+			hdr:  hdr,
+			Data: buf,
+		}
+		pf.free = func() {
+			fr.pool.Put(&buf)
+			pf.Data = nil
+		}
+		return pf, nil
+	case *http2.GoAwayFrame:
+		buf := fr.pool.Get(int(hdr.Size - 8))
+		copy(buf, f.DebugData())
+		gf := &GoAwayFrame{
+			hdr:          hdr,
+			DebugData:    buf,
+			Code:         ErrCode(f.ErrCode),
+			LastStreamID: f.LastStreamID,
+		}
+		gf.free = func() {
+			fr.pool.Put(&buf)
+			gf.DebugData = nil
+		}
+		return gf, nil
+	case *http2.WindowUpdateFrame:
+		return &WindowUpdateFrame{
+			hdr: hdr,
+			Inc: f.Increment,
+		}, nil
+	case *http2.ContinuationFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.HeaderBlockFragment())
+		return &ContinuationFrame{
+			hdr:      hdr,
+			HdrBlock: buf,
+		}, nil
+	case *http2.MetaHeadersFrame:
+		return &MetaHeadersFrame{
+			hdr:    hdr,
+			Fields: f.Fields,
+		}, nil
+	}
+
+	return nil, connError(ErrCodeProtocol)
+}
+
+func (fr *HTTP2FramerBridge) WriteData(streamID uint32, endStream bool, data ...[]byte) error {
```

Comment:
Can/Should we have a special case where we avoid the copy when `len(data) == 1`?

Reply:
This is now taken care of by the `mem` package.